### ISSUE TITLE
When consuming events don't process in parallel for the same partition

### ISSUE THREAD

**Issue by mdedetrich** [zalando-nakadi/kanadi#152]
Quoting from coworker

> If we read the batch [e1, e2, e3, e4] (one partiion)
We process events in parallel
e4 was successfully processed
e2 failed
kanadi commit cursor after processing e4
e2 is lost

The problem is that Kanadi by default commits cursors in parallel (for a single partition) rather than doing them sequentially.

**Comment by mdedetrich** [comment]
The solution for this is quite simple, we just have to wait for all of the callbacks from all of the list of events to succeed before committing the cursor (this should also be behind an option)

**Comment by mdedetrich** [comment]
Okay so I just realized what the actual issue is. Assuming you have batches b1, b2 and b3 and you commit the cursor for b3 Nakadi will actually assume that b1 and b2 are also successful. This is a problem if you commit cursors at the same time as processing the next batch, i.e. you are processing b1 and it happens to fail however at the same time it is failing you process b2 and you commit the cursor for b2 before the finish of processing of b1. Committing the cursor of b2 means that Nakadi will assume that b1 is also successful which means that Kanadi client will never receive b1 in the future, essentially losing the event.

Current behavior and proposals are detailed in the calendar below
![Kanadi Proposals](https://user-images.githubusercontent.com/2337269/87180804-826f8c00-c2e1-11ea-8399-ef58f8d406be.png)

Will also need to look into what other Nakadi clients (i.e. https://github.com/zalando-nakadi/fahrschein) do. Proposal 3 appears to  look like the best solution since it doesn't reduce the performance of processing the next batch (which can matter since you only have 60 seconds to commit a cursor when you receive it).

**Comment by loki256** [comment]
@mdedetrich I would suggest to review proposal 2 again. I think it is not a good design to postpone cursor commit for b1 (it could be that processing time for b1 after b2 was processed will be more that 60 seconds and b1 will not be committed. Maybe it is possible to introduce just a queue for the cursors. You can commit cursor as soon as it is available for commit but you need to be sure that cursors are strictly ordered in the queue and you don't have situations like : Queue(c1, c3, c2) within one partition...

**Comment by mdedetrich** [comment]
Due to Kanadi being stream based, proposal 2 is very difficult (it also creates issues due to the 60 second timeout) so I would go for proposal 3. However I don't think you need a cue, you just need to store the current latest token per partition because if you commit a cursor in Nakadi then all prevous tokens are considered submitted (not just the one for the current batch)

**Comment by halyph** [comment]
I've been using `org.zalando.kanadi.api.Subscriptions.EventCallback.successPredicateFuture` as `EventCallback` in [`org.zalando.kanadi.api.Subscriptions#eventsStreamedManaged`](https://github.com/zalando-nakadi/kanadi/blob/master/src/main/scala/org/zalando/kanadi/api/Subscriptions.scala#L1400).

Let's assume we have batches [b1, b2, b3] processed in _parallel_ for the same partition.

- b1 - Failed (cursor is **not** committed)
- b2 - Failed (cursor is **not** committed)
- b3 - Success (cursor is committed)

I.e. I lost events from **b1** and **b2**, correct? 

P.S. @mdedetrich it's almost the same as you said above, I just want to be sure I understood everything properly.

**Comment by mdedetrich** [comment]
Yes this is correct. The issue is that committing cursor for b3 automatically "commits" the cursor for b1 and b2 (even though they failed to process in your application).

I wanted to have a look at this earlier but didn't have much time (its also not a trivial problem to solve). Feel free to attempt a solution if you are up to it
